### ISSUE TITLE
 Java EMA RsslSocketChannel::fillGatheringByteArray infinite loop

### ISSUE THREAD

**Issue by xanddol** [Refinitiv/Real-Time-SDK#249]
Hello, everyone!

We have faced to the issue similar to https://github.com/Refinitiv/Real-Time-SDK/issues/88 on the latest EMA library 3.7.1.0 version

Case:
We have a 10000(thousands)+ instruments on which we try to subscribe one by one using EMA java library. So we have one Consumer per 10000 instruments. In this case during subscription we are stuck in RsslSocketChannel::fillGatheringByteArray method in infinite loop.

What we have observed during debug session:
1)All the priority queues (_hightPriorityQueue, _mediumPriorityQueue and _lowPriorityQueue) in _flushOrder array are empty. At the same time the Channel state remains Active. That's why RsslSocketChannel::fillGatheringByteArray method becomes unable to obtain a TransportBufferImpl instance. The method invocation ends up with waiting for the buffer in an infinity loop.
Question: Is it possible to fix this part of the code properly  to break from the infinite loop by some condition to avoid the loop? Like some kind of workaround
2)Actually we have tried to fix it ourselves but failed on the stacktrace's step above in RsslSocketChannel::flushInternal because of _totalBytesQueued is still > 0 and we are concerned about data loss
3)I have attached VisualVM screen to show the problem and stacktrace for one subscription
4)I think that probably looks like somewhere there is a race condition but it is quite hard to debug it and find the place because If we try to have some  delay (Thread.sleep) between one by one subscriptions the problem is hardly reproducible.
5)Out EMA confuguration for the consumer
`<Channel>
		<Name value="Channel_1"/>
		<ChannelType value="ChannelType::RSSL_SOCKET"/>
		<CompressionType value="CompressionType::None"/>
		<GuaranteedOutputBuffers value="30000"/>
		<ConnectionPingTimeout value="30000"/>
		<TcpNodelay value="1"/>
		<Host value="XXXXX"/>
		<Port value="15022"/>
</Channel>
<Consumer>
		<Name value="Consumer_1"/>
		<WarmStandbyChannelSet value="WarmStandbyChannel_1"/>
		<Dictionary value="Dictionary_2"/>
		<RequestTimeout value="60000"/>
		<XmlTraceToStdout value="0"/>
</Consumer>`

![test1](https://github.com/Refinitiv/Real-Time-SDK/assets/17405016/2024bead-cc8e-4abb-9b27-53240541bac2)

UPD: by the way the simpliest test like ex340_MP_OnStreamPost in EMA where we try to subscribe on 10k+ instruments leads to the following exception as well
```
Exception in thread "pool-2-thread-1" java.lang.AssertionError
	at com.refinitiv.eta.transport.EtaQueue.poll(EtaQueue.java:57)
	at com.refinitiv.eta.transport.RsslSocketChannel.fillGatheringByteArray(RsslSocketChannel.java:2708)
	at com.refinitiv.eta.transport.RsslSocketChannel.flush(RsslSocketChannel.java:2604)
	at com.refinitiv.eta.valueadd.reactor.Worker.processChannelFlush(Worker.java:470)
	at com.refinitiv.eta.valueadd.reactor.Worker.run(Worker.java:126)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Thanks in advance!
//Andrey

**Comment by L-Karchevska** [comment]
@xanddol Thank you for reaching out with this issue! We've created an internal Jira to investigate it.

**Comment by ayurov2** [comment]
Hi @xanddol, with the AssertionError you see while running ex340_MP_OnStreamPost, if you get a chance could you please share more specifically what changes you've made to the example? Also was this connecting to the IProvider 340 Posting Example or a different provider?

I have been investigating that Assertion issue specifically but am now looking for any connections between that error and the infinite loop you are seeing in fillGatheringByteArray, so I greatly appreciate you bringing this up and updating your post when you saw the error in the 340 example.

**Comment by xanddol** [comment]
Hi all,

thanks for the response!

I've attached simple example, actually it is just a consumer test without any posting during the test
[example.txt](https://github.com/Refinitiv/Real-Time-SDK/files/12594753/example.txt)

In the test there are an obfuscated list of all instruments and an obfuscated  service on which we are subscribed during the test.
Despite of there are 100 iteration in the test we can be faced with exceptions below even in the first iteration. 

1)
Exception in thread "pool-2-thread-1" java.lang.AssertionError
	at com.refinitiv.eta.transport.EtaQueue.poll(EtaQueue.java:57)
	at com.refinitiv.eta.transport.RsslSocketChannel.fillGatheringByteArray(RsslSocketChannel.java:2708)
	at com.refinitiv.eta.transport.RsslSocketChannel.flush(RsslSocketChannel.java:2604)
	at com.refinitiv.eta.valueadd.reactor.Worker.processChannelFlush(Worker.java:470)
	at com.refinitiv.eta.valueadd.reactor.Worker.run(Worker.java:126)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

2)The infinite loop in com.refinitiv.eta.transport.RsslSocketChannel.fillGatheringByteArray as described in the beginning of the topic

3)
```
Exception in thread "pool-3-thread-1" java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 16
	at com.refinitiv.eta.codec.Decoders.decodeFieldList(Decoders.java:2404)
	at com.refinitiv.eta.codec.FieldListImpl.decode(FieldListImpl.java:56)
	at com.refinitiv.ema.access.FieldListImpl.decode(FieldListImpl.java:307)
	at com.refinitiv.ema.access.MsgImpl.decodeAttribPayload(MsgImpl.java:317)
	at com.refinitiv.ema.access.RefreshMsgImpl.decode(RefreshMsgImpl.java:623)
	at com.refinitiv.ema.access.ItemCallbackClient.processRefreshMsg(ItemCallbackClient.java:1776)
	at com.refinitiv.ema.access.ItemCallbackClient.defaultMsgCallback(ItemCallbackClient.java:1646)
	at com.refinitiv.eta.valueadd.reactor.Reactor.sendDefaultMsgCallback(Reactor.java:2787)
	at com.refinitiv.eta.valueadd.reactor.Reactor.sendAndHandleDefaultMsgCallback(Reactor.java:2896)
	at com.refinitiv.eta.valueadd.reactor.WlItemHandler.callbackUser(WlItemHandler.java:3029)
	at com.refinitiv.eta.valueadd.reactor.WlItemHandler.readRefreshMsg(WlItemHandler.java:2209)
	at com.refinitiv.eta.valueadd.reactor.WlItemHandler.readMsg(WlItemHandler.java:1923)
	at com.refinitiv.eta.valueadd.reactor.Watchlist.readMsgWSB(Watchlist.java:312)
	at com.refinitiv.eta.valueadd.reactor.Reactor.processRwfMessage(Reactor.java:4655)
	at com.refinitiv.eta.valueadd.reactor.Reactor.performChannelRead(Reactor.java:5001)
	at com.refinitiv.eta.valueadd.reactor.Reactor.dispatchAll(Reactor.java:7641)
	at com.refinitiv.ema.access.OmmBaseImpl.rsslReactorDispatchLoop(OmmBaseImpl.java:1759)
	at com.refinitiv.ema.access.OmmBaseImpl.run(OmmBaseImpl.java:1889)
```

**Comment by ayurov2** [comment]
@xanddol Could you also share what Warmstandby mode you are running- whether it is Login or Source Directory based? If it is Source Directory based, we are curious whether one server has different services than another. Additionally, a sample source directory shared from each channel can be very helpful.

If possible, providing us the full configuration details (suppress sensitive content) for the Warmstandby setup can help us replicate the setup better, as we have not yet been able to reproduce this ourselves. 

If you prefer to share configuration directly with us, you could reach out to us via the Developer Portal (https://developers.refinitiv.com): open up a query in the Q/A section and reference this GitHub issue. You may mark your question 'private'. Or you may choose to have email contact directly with whoever responds to your query on DevPortal. Note that you do not need a service contract with us in order to create an account on the Developer Portal.

**Comment by xanddol** [comment]
Hi all,
Sorry for delay.

I have attached ema config file, please share your thoughts regarding the file. 
[ema_config.txt](https://github.com/Refinitiv/Real-Time-SDK/files/12780194/ema_config.txt)

Yes it's SERVICE_BASED and both servers have exactly the same services.

**Comment by vlevendel** [comment]
@xanddol,  Please contact us via Developer Portal so that we may provide you with instrumented code to further isolate this issue.

**Comment by vlevendel** [comment]
This is included with tag, Real-Time-SDK-2.1.3.L1. Please let us know if you have further questions.
